core/src/gl/renderState.cpp

```diff
@@ -37,13 +37,74 @@ RenderState::RenderState() {
 
 }
 
+void RenderState::flushResourceDeletion() {
+    std::lock_guard<std::mutex> guard(m_deletionListMutex);
+
+    if (m_VAODeletionList.size()) {
+        GL::deleteVertexArrays(m_VAODeletionList.size(), m_VAODeletionList.data());
+        m_VAODeletionList.clear();
+    }
+    if (m_textureDeletionList.size()) {
+        GL::deleteTextures(m_textureDeletionList.size(), m_textureDeletionList.data());
+        m_textureDeletionList.clear();
+    }
+    if (m_bufferDeletionList.size()) {
+        GL::deleteBuffers(m_bufferDeletionList.size(), m_bufferDeletionList.data());
+        m_bufferDeletionList.clear();
+    }
+    if (m_framebufferDeletionList.size()) {
+        GL::deleteFramebuffers(m_framebufferDeletionList.size(), m_framebufferDeletionList.data());
+        m_framebufferDeletionList.clear();
+    }
+    if (m_programDeletionList.size()) {
+        for (GLuint program : m_programDeletionList) {
+            GL::deleteProgram(program);
+        }
+        m_programDeletionList.clear();
+    }
+    if (m_shaderDeletionList.size()) {
+        for (GLuint shader : m_shaderDeletionList) {
+            GL::deleteShader(shader);
+        }
+        m_shaderDeletionList.clear();
+    }
+}
+
+void RenderState::queueFramebufferDeletion(GLuint framebuffer) {
+    std::lock_guard<std::mutex> guard(m_deletionListMutex);
+    m_framebufferDeletionList.push_back(framebuffer);
+}
+
+void RenderState::queueProgramDeletion(GLuint program, GLuint fragShader, GLuint vertShader) {
+    std::lock_guard<std::mutex> guard(m_deletionListMutex);
+    m_programDeletionList.push_back(program);
+    m_shaderDeletionList.push_back(fragShader);
+    m_shaderDeletionList.push_back(vertShader);
+}
+
+void RenderState::queueTextureDeletion(GLuint texture) {
+    std::lock_guard<std::mutex> guard(m_deletionListMutex);
+    m_textureDeletionList.push_back(texture);
+}
+
+void RenderState::queueVAODeletion(size_t count, GLuint* vao) {
+    std::lock_guard<std::mutex> guard(m_deletionListMutex);
+    m_VAODeletionList.insert(m_VAODeletionList.end(), vao, vao + count);
```

Comment:
I think we might benefit by doing a `reserve` here before `insert`.

Reply:
I think `count > 1` is pretty rare? I believe there is several VAOs when meshes don't fit under 65k index limit, but maybe reserve an arbitrary size at initialization if you noticed reallocs here.

Reply:
We do see a few cases when meshes are >65k index limit and hence count > 1. But yes you are right there are not many which will cause significant perf issues here! I will leave it as it is right now.